sql/firefox_desktop_visited_5_uri_mau28_by_dimensions_v1.sql

```diff
@@ -0,0 +1,52 @@
+WITH
+  _sample AS (
+  SELECT
+    submission_date,
+    -- most recent values for inclusion as output dimensions
+    days[SAFE_ORDINAL(1)] AS last_seen,
+    DATE_DIFF(submission_date, (
+      SELECT
+        -- bit-packed boolean array of days meeting the where condition
+        -- with most significant bit as the least recent day
+        SUM(1 << DATE_DIFF(submission_date, submission_date_s3, DAY))
+      FROM
+        UNNEST(days)
+      WHERE
+        -- results in NULL AS _days_visited_5_uri when condition not met by any item in days
+        scalar_parent_browser_engagement_total_uri_count_sum >= 5), DAY) AS _days_visited_5_uri
+  FROM
+    clients_monthly_v1
+  WHERE
+    submission_date = @submission_date )
+SELECT
+  submission_date,
+  last_seen.attribution.source,
+  last_seen.attribution.medium,
+  last_seen.attribution.campaign,
+  last_seen.attribution.content,
+  last_seen.country,
+  last_seen.distribution_id,
+  -- BIT_COUNT returns the number of set bits
+  -- 0x0FFFFFFF is a bit-mask for the lowest 28 bits
+  -- 0x7F is a bit-mask for the lowest 7 bits
+  -- Count the number of days in the window that a client visited 5 uri
+  BIT_COUNT(_days_visited_5_uri & 0x0FFFFFFF) AS visited_5_uri_frequency28,
+  BIT_COUNT(_days_visited_5_uri & 0x7F) AS visited_5_uri_frequency7,
```

Comment:
After reading through this several times, I _think_ I understand all the bitwise logic here, but I'm not completely sure I have missed some nuance. It's definitely clever and concise, but I do worry that it's going to be difficult even for members of our team to feel confident about updating or extending this logic correctly.

Reply:
working on refactoring this to be more readable then